### DISCUSSION TITLE
Issues Exporting IPA for App Store: Exit status 70

### DISCUSSION THREAD

**Discussion by cmmourinho** [fastlane/fastlane#27962] [Q&A]
Build on : 
Xcode 16 
Fastlane 2.225.0

my lane config : 
desc "Build IPA"
  lane :build_ipa do |options|
    BUILD_NAME = options[:build_name]
    TAG = options[:tag]
    gym(
      scheme: "test2",
      export_method: "app-store",
      export_options: {
        provisioningProfiles: {
          "com.test.test2" => "testprovisioning" 
        }
      },
      output_name: "export-test.ipa"
    )
  end
  
I’m not logged into an Apple account but have manually imported the mobile provisioning profile and .p12 certificate into my keychain and no error in scheme "test2" (xcode)
  
when i run "fastlane build_ipa --verbose"
   i found error : 
   INFO [2024-11-14 15:42:26.53]: $ /usr/bin/xcrun /opt/homebrew/Cellar/fastlane/2.225.0/libexec/gems/fastlane-2.225.0/gym/lib/assets/wrap_xcodebuild/xcbuild-safe.sh -exportArchive -exportOptionsPlist '/var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/gym_config20241114-89180-m1w29y.plist' -archivePath /Users/judi/Library/Developer/Xcode/Archives/2024-11-14/export_test\ 2024-11-14\ 15.41.29.xcarchive -exportPath '/var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/gym_output20241114-89180-dmcsae' 
INFO [2024-11-14 15:42:26.56]: ▸ rbenv detected, removing env variables
INFO [2024-11-14 15:42:26.59]: ▸ rbenv: shell integration not enabled. Run `rbenv init' for instructions.
INFO [2024-11-14 15:42:26.59]: ▸ + xcodebuild -exportArchive -exportOptionsPlist /var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/gym_config20241114-89180-m1w29y.plist -archivePath '/Users/judi/Library/Developer/Xcode/Archives/2024-11-14/export_test 2024-11-14 15.41.29.xcarchive' -exportPath /var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/gym_output20241114-89180-dmcsae
INFO [2024-11-14 15:42:27.40]: ▸ 2024-11-14 15:42:27.402 xcodebuild[97892:20359993] [MT] IDEDistribution: -[IDEDistributionLogging _createLoggingBundleAtPath:]: Created bundle at path "/var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/test2_2567-11-14_15-42-27.401.xcdistributionlogs".
INFO [2024-11-14 15:42:27.51]: ▸ 2024-11-14 15:42:27.517 xcodebuild[97892:20359993] [MT] IDEDistribution: Command line name "app-store" is deprecated. Use "app-store-connect" instead.
INFO [2024-11-14 15:42:35.76]: ▸ error: exportArchive Copy failed
INFO [2024-11-14 15:42:35.76]: ▸ ** EXPORT FAILED **
rbenv detected, removing env variables
rbenv: shell integration not enabled. Run `rbenv init' for instructions.
+ xcodebuild -exportArchive -exportOptionsPlist /var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/gym_config20241114-89180-m1w29y.plist -archivePath '/Users/judi/Library/Developer/Xcode/Archives/2024-11-14/export_test 2024-11-14 15.41.29.xcarchive' -exportPath /var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/gym_output20241114-89180-dmcsae
2024-11-14 15:42:27.402 xcodebuild[97892:20359993] [MT] IDEDistribution: -[IDEDistributionLogging _createLoggingBundleAtPath:]: Created bundle at path "/var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/test2_2567-11-14_15-42-27.401.xcdistributionlogs".
2024-11-14 15:42:27.517 xcodebuild[97892:20359993] [MT] IDEDistribution: Command line name "app-store" is deprecated. Use "app-store-connect" instead.
error: exportArchive Copy failed

** EXPORT FAILED **
ERROR [2024-11-14 15:42:35.78]: Exit status: 70

I tried exporting manually, and it worked fine.
This is my method for manual export.
Archive > Distribute > Custom > App store > Export

Has anyone experienced the same issue as me?

**Comment by bella964** [comment]
Hello,
The error you're experiencing appears to be related to the xcodebuild export phase in Fastlane. Here's a step-by-step approach to troubleshoot and resolve the issue:

Key Observations:
Deprecated app-store Option: The warning about "app-store is deprecated. Use app-store-connect instead" indicates you need to update your export_method.
Manual Export Works: Since manual export works, the issue is likely related to how Fastlane/Gym is handling your configuration or the environment setup.
exportArchive Copy failed: This often relates to issues with signing or the provisioning profile during the export step.
Steps to Resolve:
1. Update export_method
Replace "app-store" with "app-store-connect" in your gym configuration:

ruby
export_method: "app-store-connect"
2. Check Provisioning Profile and Signing
Even though you've manually added the mobile provisioning profile and .p12 certificate to your keychain:

Ensure provisioning profile matches your app identifier.
Verify Keychain Access: Open Keychain Access and ensure the .p12 certificate and private key are valid and trusted.
Use the security find-identity command to confirm the certificate is available:
bash
security find-identity -v -p codesigning
3. Debug with xcodebuild
Since manual export works, run the same xcodebuild export command that Fastlane is generating, directly in the terminal, for more detailed debugging:

bash
xcodebuild -exportArchive \
    -exportOptionsPlist /var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/gym_config20241114-89180-m1w29y.plist \
    -archivePath '/Users/judi/Library/Developer/Xcode/Archives/2024-11-14/export_test 2024-11-14 15.41.29.xcarchive' \
    -exportPath /var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/gym_output20241114-89180-dmcsae
This might reveal more specific errors not captured by Fastlane.

4. Clean Up Temp Files
Fastlane uses temporary files during the build process. Clean up the temp files and try again:

bash
rm -rf /var/folders/gg/*
5. Use Explicit Signing in gym
If the issue persists, explicitly define the code signing identity and provisioning profile in your gym configuration:

ruby
gym(
  scheme: "test2",
  export_method: "app-store-connect",
  export_options: {
    provisioningProfiles: {
      "com.test.test2" => "testprovisioning"
    }
  },
  output_name: "export-test.ipa",
  export_xcargs: "-allowProvisioningUpdates",
  codesigning_identity: "iPhone Distribution: Your Name (Team ID)"
)
If Issues Persist:
Check Fastlane Logs: Use --verbose for detailed logs.
Reinstall Fastlane: If there’s a misconfiguration or bug in Fastlane, try reinstalling/updating:
bash
gem uninstall fastlane
gem install fastlane
Validate Xcode Version: Ensure that Xcode 16 is supported by Fastlane version 2.225.0.
Best Regards
[ForAgentsOnly com
](https://www.foragents-only.com)

**Comment by styv** [comment]
I also recently got the "error: exportArchive Copy failed" error when trying to use `xcodebuild` directly (i.e., not via fastlane) for exporting from an archive. It seems to me that this was introduced after installing macOS 15.4 and Xcode 16.3, but I am not 100% sure. (I am just sure that it used to work at some point in the past…)

To investigate the issue, I looked into the temporary log folder which `xcodebuild` creates for such invocations. This folder's path is printed to stdout, and in your case it would have been "/var/folders/gg/fh9vkw857kb8zw67wfhmrsn00000gr/T/test2_2567-11-14_15-42-27.401.xcdistributionlogs" (but `xcodebuild` creates a new folder for each invocation). This folder should contain a "IDEDistributionPipeline.log" file, which in my case contained an `rsync` error towards the end of the file: "rsync: on remote machine: --extended-attributes: unknown option". According to the log file, `xcodebuild` uses the system supplied `/usr/bin/rsync`. However, typing `which rsync` in Terminal resulted in `/opt/homebrew/bin/rsync`, i.e. a version installed via Homebrew. So my best explanation is that even when using `rsync` for a local-to-local copy (as `xcodebuild` is doing here), `rsync` spawns a new server instance on the local machine based on the `rsync` it finds on the current `$PATH`. And while the sending `/usr/bin/rsync` understands the `--extended-attributes` parameter, the receiving `/opt/homebrew/bin/rsync` does not.

My solution was to just uninstall the Homebrew provided `rsync` using `brew uninstall rsync`. Then, using `xcodebuild` to export from an archive worked fine again.

**Reply by yanisinho** [reply]
@styv 

Well... @styv I'm not sure that the problem is related to fastlane itself.

I'm working on another projet (React Native) and especially on an Android app.
Absolutely no related to fastlane at all, I get the stuck with a similar issue : Wrong path selected

This command : eas build --platform android --local --non-interactive --profile staging
don't pick the right yarn version.

brew uninstall yarn fixed the issue.

So I suspect that the issue is between macOS upgrade (needed for Xcode 16.3) and Homebrew.
Need to investigate a little more.

**Comment by lhunath** [comment]
The issue is caused by `xcodebuild` with Homebrew's rsync in PATH before `/usr/bin/rsync`.

You may find that the command succeeds if you prefix it with `PATH="/usr/bin:$PATH"`

**Reply by JakobLichterfeld** [reply]
Yeah, Apple replaced rsync with openrsync, which is NOT 100% compatible, xcode does know it, flutter, fastlane and others do not

**Reply by kankaristo** [reply]
Setting `PATH` like that broke Ruby for me, so I created a new directory which contains a symlink to `/usr/bin/rsync`, and then I added that directory to be the first in `PATH`. So, something like:
```
mkdir -p rsync-path
ln -fs /usr/bin/rsync rsync-path/rsync
export PATH="$PWD/rsync-path:$PATH"
```
